hbase-server/src/test/java/org/apache/hadoop/hbase/util/TestRegionMover.java

```diff
@@ -238,4 +245,139 @@ public void testTargetServerDeadWhenLoading() throws Exception {
       assertFalse(rm.load());
     }
   }
+
+  @Test
+  public void testDecomServerExclusionWithAck() throws Exception {
+    MiniHBaseCluster cluster = TEST_UTIL.getHBaseCluster();
+    HRegionServer excludeServer = cluster.getRegionServer(1);
+    List<HRegion> regions = excludeServer.getRegions();
+    int regionsExcludeServer = excludeServer.getNumberOfOnlineRegions();
+    TEST_UTIL.getAdmin().decommissionRegionServers(
+      Collections.singletonList(excludeServer.getServerName()), false);
+
+    waitForServerDecom(excludeServer);
+
+    HRegionServer regionServer = cluster.getRegionServer(0);
+    String rsName = regionServer.getServerName().getHostname();
+    int port = regionServer.getServerName().getPort();
+    String hostname = rsName + ":" + Integer.toString(port);
+    RegionMoverBuilder rmBuilder = new RegionMoverBuilder(hostname, TEST_UTIL.getConfiguration())
+      .ack(true);
+
+    int targetServerRegions = cluster.getRegionServer(2).getRegions().size();
+    int sourceServerRegions = regionServer.getRegions().size();
+
+    try (RegionMover regionMover = rmBuilder.build()) {
+      Assert.assertTrue(regionMover.unload());
+      LOG.info("Unloading {}", hostname);
+      assertEquals(0, regionServer.getNumberOfOnlineRegions());
+      assertEquals(regionsExcludeServer, cluster.getRegionServer(1).getNumberOfOnlineRegions());
+      LOG.info("Before:" + regionsExcludeServer + " After:" +
+        cluster.getRegionServer(1).getNumberOfOnlineRegions());
+      List<HRegion> regionList = cluster.getRegionServer(1).getRegions();
+      int index = 0;
+      for (HRegion hRegion : regionList) {
+        Assert.assertEquals(hRegion, regions.get(index++));
+      }
+      Assert.assertEquals(targetServerRegions + sourceServerRegions,
+        cluster.getRegionServer(2).getNumberOfOnlineRegions());
+      Assert.assertTrue(regionMover.load());
+    }
+
+    TEST_UTIL.getAdmin().recommissionRegionServer(excludeServer.getServerName(),
+      Collections.emptyList());
+  }
+
+  private void waitForServerDecom(HRegionServer excludeServer) throws IOException {
```

Comment:
Should we wait for the decom to happen or just start the region mover tool and then finally assert that none of the regions are in the decom node - assuming the region plan was created even before the decom was completed. Else LGTM.

Reply:
> assuming the region plan was created even before the decom was completed.

In this case, `admin.listDecommissionedRegionServers()` might not return server which is going to be decommissioned right? Hence, I put wait here to ensure asynchronously, RS is in decom state. What we want to do as part of `unload()` function is to not include any server which is decommissioned by retrieving them from `admin.listDecommissionedRegionServers()`. If by any chance, while unload is busy unloading regions to many region servers, we might start decommission of some servers. So in this case, we might not be able to stop unloading regions to such nodes. or let's say we unloaded regions to all other servers and after that some were put to decommissioning, if regions are still online on that servers, we should be able to load them back. Or I am missing something?